contrib/ros/rikai/contrib/ros/convert.py

```diff
@@ -0,0 +1,348 @@
+    if matched:
+        return matched[1], matched[2]
+    return None
+
+
+def import_message_type(message_type: str) -> Type[genpy.message.Message]:
+    modules = message_type.split("/")
+    try:
+        mod = importlib.import_module(".".join(modules[:-1] + ["msg"]))
+        return getattr(mod, modules[-1])
+    except ModuleNotFoundError:
+        logging.error(
+            'Could not load ROS message "%s", '
+            "please make sure the package is installed.",
+            message_type,
+        )
+        raise
+
+
+class Converter(ABC):
+    """Converter translates a ROS message or class to another type"""
+
+    @abstractmethod
+    def is_supported(self, message_type: str) -> bool:
+        """Returns True if the type is convertable"""
+
+    @abstractmethod
+    def convert(self, message_type: str, value: genpy.message.Message):
+        pass
+
+    @abstractmethod
+    def get_value(self, value, attr_name):
+        pass
+
+    @abstractmethod
+    def build_record(self, fields):
+        pass
+
+
+class JsonConverter(Converter):
+    """Convert ROS Message to Python Dict"""
+
+    CONVERT_MAP = {
+        "bool": bool,
+        "char": int,
+        "int8": int,
+        "int16": int,
+        "int32": int,
+        "int64": int,
+        # Converting the unsigned numbers might overflow.
+        "uint8": int,
+        "uint16": int,
+        "uint32": int,
+        "uint64": int,
+        "float32": float,
+        "float64": float,
+        "string": str,
+        "time": lambda t: datetime.datetime.fromtimestamp(t.to_time()),
+        "duration": lambda dur: dur.to_sec(),
+        "byte": int,
+        "byte[]": bytearray,
+        "uint8[]": bytearray,
+    }
+
+    def is_supported(
+        self,
+        message_type: str,
+    ) -> bool:
+        if message_type in self.CONVERT_MAP:
+            return True
+
+        array_type_and_size = parse_array(message_type)
+        if array_type_and_size:
+            return self.is_supported(array_type_and_size[0])
+
+        return False
+
+    def convert(self, message_type: str, value: genpy.message.Message):
+        if message_type in self.CONVERT_MAP:
+            try:
+                return self.CONVERT_MAP[message_type](value)
+            except TypeError as type_err:
+                logging.error(
+                    "Failed to convert type=%s, value=%s, converter=%s: %s",
+                    message_type,
+                    value,
+                    self.CONVERT_MAP[message_type],
+                    type_err,
+                )
+                raise
+
+        array_type_and_size = parse_array(message_type)
+        if array_type_and_size:
+            message_type = array_type_and_size[0]
+            return [self.convert(message_type, v) for v in value]
+
+        return None
+
+    def get_value(self, value, attr_name):
+        return getattr(value, attr_name)
+
+    def array_type(self, value):
+        if value is None:
```

Comment:
Nit: `return [] if value is None else [value]` ?

Reply:
done